Feature/block screenshots

estebanmino opened the pull request:
<!--
Thanks for your contribution!

Please ensure that any applicable requirements below are satisfied before submitting this pull request. This will help ensure a quick and efficient review cycle.


Screenshots blcoked on android for

- Import wallet
- Reveal private credentials (private key and seedphrase)
- Import private key

-->

**Description**

The only thing left is to add the methods where we want to block screenshots

**Checklist**

* [ ] There is a related GitHub issue
* [ ] Tests are included if applicable
* [ ] Any added code is fully documented

**Issue**

Resolves #

estebanmino committed:
native

estebanmino committed:
test on wallet

estebanmino committed:
block in some screens

andrepimenta commented on app/components/Views/ImportPrivateKeySuccess/index.js:
```diff
@@ -90,15 +100,17 @@ class ImportPrivateKeySuccess extends PureComponent {
 	};
 
-	componentWillUnmount() {
+	componentWillUnmount = async () => {
```
`async` should not be necessary?

andrepimenta commented on app/components/Views/Wallet/index.js:
```diff
@@ -106,11 +106,11 @@ class Wallet extends PureComponent {
 	}
 
-	componentDidMount() {
+	componentDidMount = async () => {
```
`async` should not be necessary?

andrepimenta commented on app/components/Views/RevealPrivateCredential/index.js:
```diff
@@ -169,8 +181,17 @@ class RevealPrivateCredential extends PureComponent {
+		});
 	}
 
+	componentWillUnmount = async () => {
```
`async` should not be necessary?

andrepimenta reviewed (changes requested):
If this is not implemented on iOS, then the respective NativeModule is not available right? This should be checked, otherwise will crash on iOS.

Also, it seems that there is no prevention for other apps to record screen, for example. Is it not possible?

Otherwise, looking good! 👍

estebanmino commented:
@andrepimenta  good call on iOS 👌

estebanmino committed:
Merge branch 'develop' into feature/block-screenshots

estebanmino committed:
ios check

estebanmino committed:
rm asyncs

estebanmino commented:
@andrepimenta all good now, added a check for iOS. Regarding the screen recording, I added the `FLAG_SECURE` flag that blocks screenshots and recording ([related link](https://developer.android.com/reference/android/view/Display?authuser=2#FLAG_SECURE))

ibrahimtaveras00 commented:
**This works on Android on the following views:**
- import seed phrase
- security & privacy > "SHOW PRIVATE KEY"
- security & privacy > "REVEAL SEED PHRASE"

showing this message on my pixel 2 
![image](https://user-images.githubusercontent.com/41752922/80512057-f6d59880-894a-11ea-913d-9c20ca603772.png)

**Few places where this doesn't work:**
- on the "secure your wallet" flow after creating your password and then navigating to the "Your seed phrase" view 
- on the import account view when adding the private key

andrepimenta commented:
@estebanmino We should consider creating a service or helper called, for example, `PreventScreenshot` that implements empty functions for iOS, so that we don't have to duplicate code like this:
`const PreventScreenshot = Platform.OS === 'android' ? NativeModules.PreventScreenshot : null;` and `PreventScreenshot && PreventScreenshot.forbid();`

A possible implementation could be
PreventScreenshot.js
`
import { NativeModules } from react-native

const isAndroid = Platform.OS === 'android' 
export default {
forbid: isAndroid ? () => NativeModules.PreventScreenshot.forbid() : () => {},
allow: isAndroid ? () => NativeModules.PreventScreenshot.allow() : () => {}
}
`

What do you think?

estebanmino committed:
Merge branch 'develop' into feature/block-screenshots

estebanmino committed:
helper

estebanmino committed:
missing places

estebanmino commented:
@andrepimenta @ibrahimtaveras00 issues addressed 👌

andrepimenta reviewed (approved):
Great! Looks good to me 👌

estebanmino committed:
Merge branch 'develop' into feature/block-screenshots

ibrahimtaveras00 reviewed (approved):
Fixes look good on Android, QA Passed 👍

estebanmino committed:
Merge branch 'develop' into feature/block-screenshots

omnat commented:
In the latest, on iOS, I could still take screenshots on these flows:
- reveal seedphrase
- during the backup flow

From my understanding, we can not prevent the user on iOS on taking a screenshot (unlike on Android), but we can show them a "taking a screenshot is not safe to do.."
